includes/class-wc-query.php

```diff
@@ -583,6 +583,95 @@ public function get_catalog_ordering_args( $orderby = '', $order = '' ) {
 		return apply_filters( 'woocommerce_get_catalog_ordering_args', $args, $orderby, $order );
 	}
 
+	/**
+	 * Add extra clauses to the product query.
+	 *
+	 * @param array    $args Product query clauses.
+	 * @param WP_Query $wp_query The current product query.
+	 * @return array The updated product query clauses array.
+	 */
+	public function product_query_post_clauses( $args, $wp_query ) {
+		$args = $this->price_filter_post_clauses( $args, $wp_query );
+		$args = $this->filter_by_attribute_post_clauses( $args, $wp_query );
+
+		$search = $this->get_main_search_query_sql();
+		if ( $search ) {
+			$args['where'] .= ' AND ' . $search;
+		}
+
+		return $args;
+	}
+
+	/**
+	 * Add the filter by attribute clauses to the product query.
+	 *
+	 * @param array    $args Product query clauses.
+	 * @param WP_Query $wp_query The current product query.
+	 * @return array The updated product query clauses array.
+	 */
+	public function filter_by_attribute_post_clauses( $args, $wp_query ) {
+		global $wpdb;
+
+		if ( ! $wp_query->is_main_query() ) {
+			return $args;
+		}
+
+		$lookup_table_name = "{$wpdb->prefix}wc_product_attributes_lookup";
+		$clause_root       = " {$wpdb->prefix}posts.ID IN (";
+		if ( 'yes' === get_option( 'woocommerce_hide_out_of_stock_items' ) ) {
+			$in_stock_clause = ' AND in_stock = 1';
+		} else {
+			$in_stock_clause = '';
+		}
+
+		$attributes_to_filter_by = $this->get_layered_nav_chosen_attributes();
+		foreach ( $attributes_to_filter_by as $taxonomy => $data ) {
+			$all_terms                  = get_terms( $taxonomy, array( 'hide_empty' => false ) );
+			$term_ids_by_slug           = wp_list_pluck( $all_terms, 'term_id', 'slug' );
+			$term_ids_to_filter_by      = array_values( array_intersect_key( $term_ids_by_slug, array_flip( $data['terms'] ) ) );
```

Comment:
I know these are IDs but this should probably be enforcing integers here because they go right into an SQL clause without escaping/sanitization.